sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercion.scala

```diff
@@ -100,6 +101,17 @@ object TypeCoercion {
     case (_: TimestampType, _: DateType) | (_: DateType, _: TimestampType) =>
       Some(TimestampType)
 
+    case (t1 @ StructType(fields1), t2 @ StructType(fields2)) if t1.sameType(t2) =>
+      Some(StructType(fields1.zip(fields2).map { case (f1, f2) =>
+        // Since `t1.sameType(t2)` is true, two StructTypes have the same DataType
+        // except `name` (in case of `spark.sql.caseSensitive=false`) and `nullable`.
+        // - Different names: use a lower case name because findTightestCommonType is commutative.
+        // - Different nullabilities: `nullable` is true iff one of them is nullable.
+        val name = if (f1.name == f2.name) f1.name else f1.name.toLowerCase(Locale.ROOT)
+        val dataType = findTightestCommonType(f1.dataType, f2.dataType).get
+        StructField(name, dataType, nullable = f1.nullable || f2.nullable)
```

Comment:
Should we follow what we are doing for `union`/`except`/`intersect`? Always pick the name of the head one?

Reply:
See the example, 
```Scala
      sql("SELECT 1 as a UNION ALL (SELECT 1 as A)").show()
      sql("SELECT 1 as A UNION ALL (SELECT 1 as a)").show()
```

Reply:
This PR works as you want. This function is used to compare the equality only. BTW, for this function, it should should lower (or upper case) because it should be commutative.
```
scala> sql("SELECT struct(1 a) UNION ALL (SELECT struct(2 A))").printSchema
root
 |-- named_struct(a, 1 AS `a`): struct (nullable = false)
 |    |-- a: integer (nullable = false)

scala> sql("SELECT struct(1 A) UNION ALL (SELECT struct(2 a))").printSchema
root
 |-- named_struct(A, 1 AS `A`): struct (nullable = false)
 |    |-- A: integer (nullable = false)
```

Reply:
```
val name = if (f1.name == f2.name) f1.name else f1.name.toLowerCase(Locale.ROOT)
```

The above code changes the case, right?

Reply:
Sure, right. It's for commutativity.

Reply:
Please see [TypeCoercionSuite.checkWidenType](https://github.com/apache/spark/blob/master/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/TypeCoercionSuite.scala#L130-L142).

In order to use the first type name, we need to loosen this test helper function and to break the existing commutative assumption. I'm ok for that if you want.